### PR TITLE
[superseded by #3451][RFC, l2arc accounting & checksum issue] l2arc-write-target-size.diff , cumulative patch

### PR DESCRIPTION
**Pull request by kernelOfTruth**
l2arc-write-target-size.diff
suggested by Andriy Gapon ( @avg-I ) on the FreeBSD mailing list as a fix

which is a cumulation of the following patches:

consistently use asize for allocated size
>In the context of l2arc code we not need to explicitly use psize at all.
Logical size is used to track actual buffer size.  Allocated size is
used to track disk space and offsets.  Physical (compressed) size is
not needed.

l2arc_compress_buf: zero buffer tail only if compression succeeds
>The compression is considered successful if the size of compressed
data is less than the original data size after rounding up to
into account the vdev ashift.  It does not make sense to have the
data compressed if all savings are lost to alignment.
>
>This change also prevents a potential memory corruption if the
rounded compressed size is greater than the original size as
previously we would zero out the buffer tail before checking for
the condition.
>
>A possible corruption scenario.
>Original size is 10KB, so the data is stored in a 10KB buffer.
>Thus, a 10KB buffer would also be allocated for the compressed data.
Let's assume that the data gets compressed to 8KB + 1B size
(reduction by almost 20%).  Let's further assume that the ashift is 12.
So the compressed size would be rounded up to 12KB.  The code would
zeroe out the trailing part of the buffer, so it would overwrite 2KB
beyond the end of the buffer.

l2arc_compress_buf: remove redundant check of an unsigned value

l2arc: restore correct rounding up of asize of compressed data
>This rounding up was lost in a mismerge of illumos code.
>See
>  r268075 MFV r267565
>  f95fd16
>  illumos/illumos-gate@5d7b4d4
>It was originally introduced in r256889, 1c55b38

Revert "Transform the I/O when vdev_physical_ashift is greater than"
>This reverts commit c9e947a.
>
>Physical zio must not be silently modified.

account for ashift when choosing buffers to be written to l2arc device
>If we don't account for that, then we might end up overwriting disk
area of buffers that have not been evicted yet, because l2arc_evict
operates in terms of disk addresses.

References:
https://github.com/freebsd/freebsd/compare/master...avg-I:review/l2arc-write-target-size
http://lists.freebsd.org/pipermail/freebsd-fs/2014-October/020242.html
[needs a more extensive list]

Ported-by:
kernelOfTruth kernelOfTruth@gmail.com

Fixes (supposedly):
zfsonlinux#3114
zfsonlinux#3400
[a potentially several others]

### CODE REVIEW

**Comment by sempervictus** [comment]
Thanks as always @kernelOfTruth.
It builds, i can tell you that much.
Aside from my personal host, i'm having some trouble testing this as it collides with abd_next and #3216 pretty badly.
Any chance i could ask you to do a new conjoined PR consisting of abd_next, #3216, and this PR? I could then put it into the full testing rotation @ SVIT without breaking our current evals of #3115 or the pretty much required ABD stack for our workhorses and production systems.

**Comment by kernelOfTruth** [comment]
@sempervictus pushed to #3216 

I'll push the change to abd_next another time

Take the buildbot results with a grain of salt since they, according to @behlendorf , currently don't cover L2ARC testing

**Comment by sempervictus** [comment]
Running the following stack on my test host presently:
```
  * origin/pr/3308
  ** 3.12 compat, 4.0 compat for super_operations shrinker callbacks
  * origin/pr/3421
  ** Kernel 4.1 changes
  * origin/pr/3396
  ** Illumos 5269 - zpool import slow
  * origin/pr/2012
  ** Add option to zpool status to print guids
  * origin/pr/2557
  ** Added version command to zpool program
  * origin/pr/2668
  ** Allow for "zfs receive" to skip existing snapshots
  * origin/pr/2784
  ** Illumos #4950 files sometimes can't be removed from a full filesystem
  * origin/pr/3166
  ** Make linking with and finding libblkid required
  * origin/pr/3169
  ** Add dfree_zfs for changing how Samba reports space
  * origin/pr/3307
  ** zfsdev_getminor() should check for invalid file handles
  * remotes/kernelOfTruth/tuxoko_zfs/abd_next_19.05.2015+3433
  ** l2arc-write-target-size.diff
  * master @ 65037d9b25c2bfa98d0aa5c9e34678127c03b345
```
I've wrapped around L2ARC once already, and its working properly.
I skimmed the PR, but will need to look at the sources in context - i cant glean from the changes why in the world this would work suddenly. Still pushing data for a second wraparound, should know by morning.

**Comment by sempervictus** [comment]
Looks like the L2ARC is holding up. However, ARC itself isn't doing so well:

```
    time  read  miss  miss%  dmis  dm%  pmis  pm%  mmis  mm%  arcsz     c  
13:32:48     0     0      0     0    0     0    0     0    0   6.1G  2.3G  
root@zfs-test00:/tmp# cat /sys/module/zfs/parameters/zfs_arc_max
4294967296
root@zfs-test00:/tmp# cat /proc/spl/kstat/zfs/arcstats | grep size
size                            4    6549124784
hdr_size                        4    69960
data_size                       4    0
meta_size                       4    4935680
other_size                      4    1342728
anon_size                       4    2115584
mru_size                        4    808960
mru_ghost_size                  4    0
mfu_size                        4    2011136
mfu_ghost_size                  4    0
l2_size                         4    153742000128
l2_asize                        4    24358981632
l2_hdr_size                     4    6542776416
duplicate_buffers_size          4    0
```

Looks like the L2 headers are overrunning the ARC cache limits and OOMing the VMs running from the local ZVOLs. Not ideal, but at least the L2ARCs aren't showing insane capacities anymore.

**Comment by kernelOfTruth** [comment]
@sempervictus thanks *_a lot_* for your report !

ok, so now that L2ARC is working as intended ™ (I can't see where that regression might have come from - unless L2ARC didn't work properly or at all prior to that patchset)

it's time to fix it up - then ZFS should be ready to take off =)


That huge memory consumption was supposed to be fixed by "Illumos - 5408 managing ZFS cache devices requires lots of RAM"

From what branch are those results, btw ?

**Comment by sempervictus** [comment]
Thanks @kernelOfTruth, this is starting to come together.
I have noticed something disturbing from the patch stack however - l2arc_adapt is consuming more CPU than any of the VMs except one (large JVM procs and MySQL atop that one). The IOWait is growing commensurately with the CPU usage by the adapt thread, with 20% on the physical host registering as 70% (!!!) in the VMs. The heavy VM with the DB and JVM procs died from timeouts :(.

I'll build out #3216, but with large block having been adopted one some hosts, i can't test without destroying their pools.

I noticed @dweeezil pushed more commits to #3115, hopefully we can get that rebased against the large block feature and abd_next to create a stable stack which can be tested across current ZFS hosts.

Thanks again, will post more results as i get 'em.

**Comment by sempervictus** [comment]
Built a basic stack off of #3216:
```
  * origin/pr/2557
  ** Added version command to zpool program
  * origin/pr/3169
  ** Add dfree_zfs for changing how Samba reports space
  * origin/pr/3166
  ** Make linking with and finding libblkid required
  * origin/pr/3307
  ** zfsdev_getminor() should check for invalid file handles
  * origin/pr/2668
  ** Allow for "zfs receive" to skip existing snapshots
  * origin/pr/2012
  ** Add option to zpool status to print guids
  * master @ 65037d9b25c2bfa98d0aa5c9e34678127c03b345
```
Results are promising:
```
size                            4    4306210160
hdr_size                        4    284803296
data_size                       4    2625819648
meta_size                       4    103254016
other_size                      4    111221872
anon_size                       4    327857152
mru_size                        4    41394176
mru_ghost_size                  4    4263519744
mfu_size                        4    2359822336
mfu_ghost_size                  4    21323776
l2_size                         4    52562030592
l2_asize                        4    9820920320
l2_hdr_size                     4    1181111328
duplicate_buffers_size          4    0
root@zfs-test00:~# arcstat 
    time  read  miss  miss%  dmis  dm%  pmis  pm%  mmis  mm%  arcsz     c  
00:24:11     0     0      0     0    0     0    0     0    0   4.0G  4.0G
```
Will take a while to wrap around the L2ARC. I'll push a deep `find` or something to try and force more data into the ARC. If this holds up on the bonnie VM through the night i'll see about pushing to one of our SCST hosts since a couple are running sans L2ARC right now.

**Comment by avg-I** [comment]
I think that the changes to l2arc_write_buffers() might not be entirely correct.
The version of the patch found in this [review request](https://reviews.csiden.org/r/112/diff/1/#index_header) is probably more correct.
Note `arcstat_l2_asize` and `vdev_space_update` deltas.

The problem is that elsewhere throughout the ARC code those statistics are updated based on `b_asize` from the L2 header. `b_asize` is not entirely correct name because its value does not account for ashift if the buffer is not compressed.

Changing `b_asize` to really be *allocated* size (rather than *physical* size as it is now) is a lot of work for which I didn't have time.
Another possibility is to change all the places were `arcstat_l2_asize` and `vdev_space_update` are updated, so that `b_asize` is passed through `vdev_psize_to_asize()`. I didn't do that either.
Ultimately, I decided to keep the current meaning of those two statistics. That meaning is not correct, but it's historically acceptable.

To summarize: the current code in this pull request updates `arcstat_l2_asize` and `vdev_space_update` inconsistently, so the stats would grow over time. I am not sure what would be a good test to observe that.

**Comment by kernelOfTruth** [comment]
@avg-I Thank you very much for your detailed explanation !

I've pushed a pull-request to let the build-bots crunch on it & to have the commit available for testing

Now waiting to hear back from the upstream review ... 

via @avg-I :
> Is it possible to make me an owner of this request?
https://reviews.csiden.org/r/112/

CC: @skiselkov , @ahrens , @buffyg

**Comment by kernelOfTruth** [comment]
superseded by 

https://github.com/zfsonlinux/zfs/pull/3451
https://github.com/zfsonlinux/zfs/pull/3491
